sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CollapseProjectSuite.scala

```diff
@@ -266,4 +267,35 @@ class CollapseProjectSuite extends PlanTest {
     val expected = relation.select(($"a" + $"b").as("C")).analyze
     comparePlans(optimized, expected)
   }
+
+  test("ES-1102888: collapse project duplicating aggregate expressions in UDF") {
+    withSQLConf(SQLConf.AVOID_COLLAPSE_UDF_WITH_EXPENSIVE_EXPR.key -> "true") {
+      val pythonUdf = (e: Expression) => {
+        PythonUDF("udf", null, ArrayType(IntegerType), Seq(e), 0, udfDeterministic = true)
+      }
+
+      val query = testRelation
+        .groupBy($"a")(collectList($"b").as("l1"))
+        .select(pythonUdf($"l1").as("l2"))
+        .select(CreateArray(Seq(
+          GetArrayItem($"l2", 0),
+          GetArrayItem($"l2", 1),
+          GetArrayItem($"l2", 2),
+          GetArrayItem($"l2", 3)
+        )))
+        .analyze
+
+      val optimized = Optimize.execute(query)
+      val expected = testRelation
+        .groupBy($"a")(pythonUdf(collectList($"b")).as("l2"))
```

Comment:
shall we add a negative test which we don't end up collapsing the projects?

Reply:
this technically is also a negative test, since before this change, this would've been collapsed into a single `Aggregate` node (without the `Project`)